File: sdk/runanywhere-swift/Sources/RunAnywhere/Features/Cloud/RoutingTelemetry.swift
//
//  RoutingTelemetry.swift
//  RunAnywhere SDK
//
//  Telemetry events for routing decisions and cloud usage.
//

import Foundation

// MARK: - Routing Event

/// Event emitted when a routing decision is made.
///
/// Subscribe via EventBus to track routing patterns:
/// ```swift
/// EventBus.shared.events(for: .llm)
///     .compactMap { $0 as? RoutingEvent }
///     .sink { event in
///         print("Routed to \(event.executionTarget), confidence: \(event.confidence)")
///     }
/// ```
public struct RoutingEvent: SDKEvent {
    public let id: String = UUID().uuidString
    public let type: String = "routing.decision"
    public let category: EventCategory = .llm
    public let timestamp: Date = Date()
    public let sessionId: String? = nil
    public let destination: EventDestination = .all

    // MARK: - Routing-Specific Properties

    /// The routing mode that was configured
    public let routingMode: RoutingMode

    /// Where inference was actually executed
    public let executionTarget: ExecutionTarget

    /// On-device confidence score (0.0-1.0)
    public let confidence: Float

    /// Whether cloud handoff was triggered
    public let cloudHandoffTriggered: Bool

    /// Reason for handoff
    public let handoffReason: HandoffReason

    /// Cloud provider used (nil if on-device only)
    public let cloudProviderId: String?

    /// Cloud model used (nil if on-device only)
    public let cloudModel: String?

    /// Total latency in milliseconds
    public let latencyMs: Double

    /// Estimated cloud cost in USD (nil for on-device)
    public let estimatedCostUSD: Double?

    // MARK: - SDKEvent

    public var properties: [String: String] {
        var props: [String: String] = [
            "routing_mode": routingMode.rawValue,
            "execution_target": executionTarget.rawValue,
            "confidence": String(format: "%.4f", confidence),
            "cloud_handoff": String(cloudHandoffTriggered),
            "handoff_reason": String(handoffReason.rawValue),
            "latency_ms": String(format: "%.1f", latencyMs),
        ]
        if let id = cloudProviderId { props["cloud_provider_id"] = id }
        if let model = cloudModel { props["cloud_model"] = model }
        if let cost = estimatedCostUSD { props["estimated_cost_usd"] = String(format: "%.6f", cost) }
        return props
    }
}

// MARK: - Cost Event

/// Event emitted when a cloud request incurs cost.
public struct CloudCostEvent: SDKEvent {
    public let id: String = UUID().uuidString
    public let type: String = "cloud.cost"
    public let category: EventCategory = .llm
    public let timestamp: Date = Date()
    public let sessionId: String? = nil
    public let destination: EventDestination = .analyticsOnly

    /// Provider that incurred the cost
    public let providerId: String

    /// Input tokens
    public let inputTokens: Int

    /// Output tokens
    public let outputTokens: Int

    /// Estimated cost in USD
    public let costUSD: Double

    /// Cumulative total after this request
    public let cumulativeTotalUSD: Double

    public var properties: [String: String] {
        [
            "provider_id": providerId,
            "input_tokens": String(inputTokens),
            "output_tokens": String(outputTokens),
            "cost_usd": String(format: "%.6f", costUSD),
            "cumulative_total_usd": String(format: "%.6f", cumulativeTotalUSD),
        ]
    }
}

// MARK: - Provider Failover Event

/// Event emitted when a provider failover occurs.
public struct ProviderFailoverEvent: SDKEvent {
    public let id: String = UUID().uuidString
    public let type: String = "cloud.provider_failover"
    public let category: EventCategory = .llm
    public let timestamp: Date = Date()
    public let sessionId: String? = nil
    public let destination: EventDestination = .all

    /// Provider that failed
    public let failedProviderId: String

    /// Provider that was used as fallback
    public let fallbackProviderId: String?

    /// Error from the failed provider
    public let failureReason: String

    public var properties: [String: String] {
        var props: [String: String] = [
            "failed_provider_id": failedProviderId,
            "failure_reason": failureReason,
        ]
        if let fallback = fallbackProviderId {
            props["fallback_provider_id"] = fallback
        }
        return props
    }
}

// MARK: - Latency Timeout Event

/// Event emitted when a latency timeout triggers cloud fallback.
public struct LatencyTimeoutEvent: SDKEvent {
    public let id: String = UUID().uuidString
    public let type: String = "routing.latency_timeout"
    public let category: EventCategory = .llm
    public let timestamp: Date = Date()
    public let sessionId: String? = nil
    public let destination: EventDestination = .all

    /// Maximum allowed latency (ms)
    public let maxLatencyMs: UInt32

    /// Actual elapsed time before timeout (ms)
    public let actualLatencyMs: Double

    public var properties: [String: String] {
        [
            "max_latency_ms": String(maxLatencyMs),
            "actual_latency_ms": String(format: "%.1f", actualLatencyMs),
        ]
    }
}

File: sdk/runanywhere-flutter/packages/runanywhere/lib/features/cloud/routing_telemetry.dart
/// Routing Telemetry
///
/// Telemetry events for routing decisions and cloud usage.
/// Mirrors Swift RoutingTelemetry.swift from Features/Cloud/RoutingTelemetry.swift
library routing_telemetry;

import 'package:runanywhere/features/cloud/cloud_types.dart';
import 'package:runanywhere/public/events/sdk_event.dart';

// MARK: - Routing Event

/// Event emitted when a routing decision is made.
///
/// Subscribe via EventBus to track routing patterns:
/// ```dart
/// EventBus.shared.allEvents
///     .where((e) => e is RoutingTelemetryEvent)
///     .cast<RoutingTelemetryEvent>()
///     .listen((event) {
///       print('Routed to ${event.executionTarget}, confidence: ${event.confidence}');
///     });
/// ```
///
/// Matches Swift RoutingEvent struct.
class RoutingTelemetryEvent with SDKEventDefaults {
  @override
  String get type => 'routing.decision';

  @override
  EventCategory get category => EventCategory.llm;

  @override
  EventDestination get destination => EventDestination.all;

  // MARK: - Routing-Specific Properties

  /// The routing mode that was configured
  final RoutingMode routingMode;

  /// Where inference was actually executed
  final ExecutionTarget executionTarget;

  /// On-device confidence score (0.0-1.0)
  final double confidence;

  /// Whether cloud handoff was triggered
  final bool cloudHandoffTriggered;

  /// Reason for handoff
  final HandoffReason handoffReason;

  /// Cloud provider used (null if on-device only)
  final String? cloudProviderId;

  /// Cloud model used (null if on-device only)
  final String? cloudModel;

  /// Total latency in milliseconds
  final double latencyMs;

  /// Estimated cloud cost in USD (null for on-device)
  final double? estimatedCostUSD;

  RoutingTelemetryEvent({
    required this.routingMode,
    required this.executionTarget,
    required this.confidence,
    required this.cloudHandoffTriggered,
    required this.handoffReason,
    this.cloudProviderId,
    this.cloudModel,
    required this.latencyMs,
    this.estimatedCostUSD,
  });

  @override
  Map<String, String> get properties {
    final props = <String, String>{
      'routing_mode': routingMode.value,
      'execution_target': executionTarget.value,
      'confidence': confidence.toStringAsFixed(4),
      'cloud_handoff': cloudHandoffTriggered.toString(),
      'handoff_reason': handoffReason.code.toString(),
      'latency_ms': latencyMs.toStringAsFixed(1),
    };
    if (cloudProviderId != null) {
      props['cloud_provider_id'] = cloudProviderId!;
    }
    if (cloudModel != null) {
      props['cloud_model'] = cloudModel!;
    }
    if (estimatedCostUSD != null) {
      props['estimated_cost_usd'] = estimatedCostUSD!.toStringAsFixed(6);
    }
    return props;
  }
}

// MARK: - Cost Event

/// Event emitted when a cloud request incurs cost.
///
/// Matches Swift CloudCostEvent struct.
class CloudCostTelemetryEvent with SDKEventDefaults {
  @override
  String get type => 'cloud.cost';

  @override
  EventCategory get category => EventCategory.llm;

  @override
  EventDestination get destination => EventDestination.analyticsOnly;

  /// Provider that incurred the cost
  final String providerId;

  /// Input tokens
  final int inputTokens;

  /// Output tokens
  final int outputTokens;

  /// Estimated cost in USD
  final double costUSD;

  /// Cumulative total after this request
  final double cumulativeTotalUSD;

  CloudCostTelemetryEvent({
    required this.providerId,
    required this.inputTokens,
    required this.outputTokens,
    required this.costUSD,
    required this.cumulativeTotalUSD,
  });

  @override
  Map<String, String> get properties => {
        'provider_id': providerId,
        'input_tokens': inputTokens.toString(),
        'output_tokens': outputTokens.toString(),
        'cost_usd': costUSD.toStringAsFixed(6),
        'cumulative_total_usd': cumulativeTotalUSD.toStringAsFixed(6),
      };
}

// MARK: - Provider Failover Event

/// Event emitted when a provider failover occurs.
///
/// Matches Swift ProviderFailoverEvent struct.
class ProviderFailoverTelemetryEvent with SDKEventDefaults {
  @override
  String get type => 'cloud.provider_failover';

  @override
  EventCategory get category => EventCategory.llm;

  @override
  EventDestination get destination => EventDestination.all;

  /// Provider that failed
  final String failedProviderId;

  /// Provider that was used as fallback
  final String? fallbackProviderId;

  /// Error from the failed provider
  final String failureReason;

  ProviderFailoverTelemetryEvent({
    required this.failedProviderId,
    this.fallbackProviderId,
    required this.failureReason,
  });

  @override
  Map<String, String> get properties {
    final props = <String, String>{
      'failed_provider_id': failedProviderId,
      'failure_reason': failureReason,
    };
    if (fallbackProviderId != null) {
      props['fallback_provider_id'] = fallbackProviderId!;
    }
    return props;
  }
}

// MARK: - Latency Timeout Event

/// Event emitted when a latency timeout triggers cloud fallback.
///
/// Matches Swift LatencyTimeoutEvent struct.
class LatencyTimeoutTelemetryEvent with SDKEventDefaults {
  @override
  String get type => 'routing.latency_timeout';

  @override
  EventCategory get category => EventCategory.llm;

  @override
  EventDestination get destination => EventDestination.all;

  /// Maximum allowed latency (ms)
  final int maxLatencyMs;

  /// Actual elapsed time before timeout (ms)
  final double actualLatencyMs;

  LatencyTimeoutTelemetryEvent({
    required this.maxLatencyMs,
    required this.actualLatencyMs,
  });

  @override
  Map<String, String> get properties => {
        'max_latency_ms': maxLatencyMs.toString(),
        'actual_latency_ms': actualLatencyMs.toStringAsFixed(1),
      };
}
